client/src/sge/GraphicsGeometry.cpp

```diff
@@ -329,6 +443,125 @@ namespace sge {
+        for (unsigned int i = 0; i < animation.mNumChannels; i++) {
+            std::string nodeName = animation.mChannels[i]->mNodeName.C_Str();
+            // Load missing bones in animation
+            if (!boneMap.count(nodeName)) {
+                boneMap[nodeName] = boneMap.size();
+                bones.offsetMatrices.push_back(glm::mat4(1)); // To maintain invariant that this vector can always be indexed by bone id, tho this matrix (should) never be used
+                numBones++;
+            }
+            int id = boneMap[nodeName];
+            cur.channels[id] = BonePose(*animation.mChannels[i], id);
+        }
+        return cur;
+    }
+
+    /**
+     * PRECONDITION: All meshes already loaded (and by extension, all bones directly attached to vertices have been loaded)
+     *
+     * Loads the model's bone hierarchy and load in any missing bones
+     * @param root aiNode to build from
+     * @return Resulting tree translated to ModelComposite::BoneNode's
+     */
+    ModelComposite::BoneNode ModelComposite::buildBoneHierarchy(aiNode *root) {
+        BoneNode cur;
+        std::string boneName = root->mName.C_Str();
+        // Assimp is dumb and adds a bunch of extra nodes in the hierarchy tree with identity matrix relative transformations :/
+        int boneId = -1;
+        if (boneMap.count(boneName)) {
+            boneId = boneMap[boneName];
+        }
+        cur.relativeTransform = assimpToGlmMat4(root->mTransformation);
+        cur.id = boneId;
+        for (unsigned int i = 0; i < root->mNumChildren; i++) {
+            cur.children.push_back(buildBoneHierarchy(root->mChildren[i]));
+            // Delete unnecessary nodes
+            if (cur.children.back().id == -1 && cur.children.back().children.empty()) {
+                cur.children.pop_back();
+            }
+        }
+
+        return cur;
+    }
+
+    /**
+     * Return model pose for a given animation at a given timestamp
+     * @param animationId Model's animation identifier
+     * @param time Timestamp to retrieve pose
+     * @return Model's pose at given timestamp
+     */
+    void ModelComposite::animationPose(int animationId, float time, ModelPose& outputModelPose) {
+        assert(animationId >= -1  && animationId < animations.size() && animated);
+        if (animationId == -1) {
+            animationId = animationWhenStill;
+        }
+        Animation anim = animations[animationId];
+        glm::mat4 accumulator(1);
+        // Recursively construct final transformation matrices for each bone
+        recursePose(outputModelPose, anim, time, accumulator, bones.root);
+    }
+
+    ModelPose ModelComposite::emptyModelPose() {
+        ModelPose pose(MAX_BONES, glm::mat4(1));
+        return pose;
+    }
+
+    float ModelComposite::timeToAnimationTick(long long milliseconds, int animationId) {
+        assert(animationId >= 0 && animationId < animations.size() && animated);
```

Comment:
Why do we assert animationId >= 0 if there is a check on the next line for animationId == -1?

Reply:
changed to check for animation id >= -1